### PR TITLE
fix #700

### LINKED ISSUE #700: Installation issue under linux
Hello,
stlink installation under Linux (Debian 9) is broken since modifications for #695: libstlink.so (and others) installs in /usr/local/lib/lib instead of /usr/local/lib
It seems that the error comes from these instructions:
 if(IS_DIRECTORY CMAKE_LIBRARY_PATH)
 else()
    set(CMAKE_LIBRARY_PATH "lib")
 endif()
If I change it to 
 if(IS_DIRECTORY CMAKE_LIBRARY_PATH)
 else()
    set(CMAKE_LIBRARY_PATH "")
 endif()
then it works again.
I can create a PR if you want, but I cannot test the modification under other OS.

### PR DESCRIPTION
**Pull request by slyshykO**
Use CMAKE_INSTALL_PREFIX for install paths

Need to approve from @cedricboudinet and @Vascom

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,17 +5,7 @@ set(PROJECT_DESCRIPTION "Open source version of the STMicroelectronics Stlink To
...
-else()
-    set(CMAKE_LIBRARY_PATH "lib")
-endif()
-
-if (WIN32)
-    set(STLINK_LIBRARY_PATH ${CMAKE_LIBRARY_PATH} CACHE PATH "Target lib directory")
-else()
-    set(STLINK_LIBRARY_PATH "${LIB_INSTALL_DIR}/${CMAKE_LIBRARY_PATH}" CACHE PATH "Target lib directory")
-endif(WIN32)
+set(STLINK_LIBRARY_PATH "${CMAKE_INSTALL_PREFIX}/${LIB_INSTALL_DIR}" CACHE PATH "Target lib directory")
```
Comment (Vascom) [inline CMakeLists.txt]: Seems that LIB_INSTALL_DIR already /usr/lib64 and CMAKE_INSTALL_PREFIX = /usr.
So lib installed to /usr/usr/lib64/
Need to remove CMAKE_INSTALL_PREFIX from this line.

**Comment by slyshykO** [comment]
Hi, Vasilii.  Did you remove cmake cache before the test?

пт, 30 мар. 2018 г., 9:41 Vasiliy Glazov <notifications@github.com>:

**Comment by Vascom** [comment]
I don't know how to remove it. But on clean installed build environment it has same error:

> Installing: /builddir/build/BUILDROOT/stlink-1.5.0-1.fc27.x86_64/usr/usr/lib64/libstlink.so.1.5.0

Full build log https://kojipkgs.fedoraproject.org//work/tasks/5046/26055046/build.log
https://koji.fedoraproject.org/koji/taskinfo?taskID=26055046

**Comment by Vascom** [comment]
Not work.
In CMakeCache.txt it has lines:
```
//No help, variable specified on the command line.
LIB_INSTALL_DIR:PATH=/usr/lib64

//Target include directory
STLINK_INCLUDE_PATH:PATH=/usr/usr/include

//Target lib directory
STLINK_LIBRARY_PATH:PATH=/usr/usr/lib64
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,18 +4,22 @@ project(stlink C)
 set(PROJECT_DESCRIPTION "Open source version of the STMicroelectronics Stlink Tools")
 set(STLINK_UDEV_RULES_DIR "/etc/udev/rules.d" CACHE PATH "Udev rules directory")
 set(STLINK_MODPROBED_DIR "/etc/modprobe.d" CACHE PATH "modprobe.d directory")
-set(LIB_INSTALL_DIR "lib" CACHE PATH "Main library directory")
 
-if(IS_DIRECTORY CMAKE_LIBRARY_PATH)
```
Comment (Vascom) [inline CMakeLists.txt]: ${LIB_INSTALL_DIR}

**Comment by slyshykO** [comment]
Vasiliy. You pass to cmake those variables from a command line:
CMAKE_INSTALL_PREFIX:PATH=/usr
INCLUDE_INSTALL_DIR:PATH=/usr/include
LIB_INSTALL_DIR:PATH=/usr/lib64

Can **you** change that to:
CMAKE_INSTALL_PREFIX:PATH=/usr
INCLUDE_INSTALL_DIR:PATH=include
LIB_INSTALL_DIR:PATH=lib64

And I, respectively, will set in cmake only:
set(STLINK_LIBRARY_PATH "${CMAKE_INSTALL_PREFIX}/${LIB_INSTALL_DIR}" CACHE PATH "Target lib directory")
set(STLINK_INCLUDE_PATH "${CMAKE_INSTALL_PREFIX}/${INCLUDE_INSTALL_DIR}" CACHE PATH "Target include directory")

without any checks

File: CMakeLists.txt [hunk 3]
```diff
@@ -4,18 +4,22 @@ project(stlink C)
...
+    set(STLINK_LIBRARY_PATH "${LIB_INSTALL_DIR}" CACHE PATH "Target lib directory")
 else()
-    set(CMAKE_LIBRARY_PATH "lib")
+    set(LIB_INSTALL_DIR "lib" CACHE PATH "Main library directory")
+    set(STLINK_LIBRARY_PATH "${CMAKE_INSTALL_PREFIX}/${LIB_INSTALL_DIR}" CACHE PATH "Target lib directory")
 endif()
 
-if (WIN32)
-    set(STLINK_LIBRARY_PATH ${CMAKE_LIBRARY_PATH} CACHE PATH "Target lib directory")
+if( IS_DIRECTORY INCLUDE_INSTALL_DIR)
```
Comment (Vascom) [inline CMakeLists.txt]: ${INCLUDE_INSTALL_DIR}

**Comment by Vascom** [comment]
> Can you change that to:
I cant change it. You need just correct variables in "if("

**Comment by Vascom** [comment]
Now all OK.

**Comment by slyshykO** [comment]
Will wait for approving from Debian guy.

**Comment by xor-gate** [comment]
@bluca and @zoobab are the guys who did some debian packaging and could probably verify this works also.

**Comment by bluca** [comment]
I'm not a CMake expert (in fact I have an intense dislike for it...), but something is not right:
as of https://github.com/texane/stlink/pull/683 everything was working fine for me, with the following config:

```
cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_LIBRARY_PATH=x86_64-linux-gnu -DSTLINK_UDEV_RULES_DIR=/lib/udev/rules.d
```

I got what I needed:
```
$ dpkg -c libstlink1_1.5.0+ds-1_amd64.deb 
drwxr-xr-x root/root         0 2018-03-16 16:56 ./
drwxr-xr-x root/root         0 2018-03-16 16:56 ./usr/
drwxr-xr-x root/root         0 2018-03-16 16:56 ./usr/lib/
drwxr-xr-x root/root         0 2018-03-16 16:56 ./usr/lib/x86_64-linux-gnu/
-rw-r--r-- root/root     80256 2018-03-16 16:56 ./usr/lib/x86_64-linux-gnu/libstlink.so.1.5.0
drwxr-xr-x root/root         0 2018-03-16 16:56 ./usr/share/
drwxr-xr-x root/root         0 2018-03-16 16:56 ./usr/share/doc/
drwxr-xr-x root/root         0 2018-03-16 16:56 ./usr/share/doc/libstlink1/
-rw-r--r-- root/root      2175 2018-03-16 16:56 ./usr/share/doc/libstlink1/changelog.Debian.gz
-rw-r--r-- root/root      2751 2018-02-19 19:00 ./usr/share/doc/libstlink1/changelog.gz
-rw-r--r-- root/root      9922 2018-03-16 16:56 ./usr/share/doc/libstlink1/copyright
lrwxrwxrwx root/root         0 2018-03-16 16:56 ./usr/lib/x86_64-linux-gnu/libstlink.so.1 -> libstlink.so.1.5.0
```

Building from this PR this doesn't work anymore, CMAKE_LIBRARY_PATH is ignored, this is where libraries are installed now:

```
$ find debian/tmp/usr/lib/
debian/tmp/usr/lib/
debian/tmp/usr/lib/libstlink.so.1
debian/tmp/usr/lib/libstlink.so.1.4.0
debian/tmp/usr/lib/libstlink.so
debian/tmp/usr/lib/pkgconfig
debian/tmp/usr/lib/pkgconfig/stlink.pc
debian/tmp/usr/lib/libstlink.a
```

Setting LIB_INSTALL_DIR:

```
cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DLIB_INSTALL_DIR=debian/tmp/usr/lib/x86_64-linux-gnu -DSTLINK_UDEV_RULES_DIR=/lib/udev/rules.d
```

Makes it even worse, not sure where that path is coming from:

```
$ find debian/tmp/usr/home
debian/tmp/usr/home
debian/tmp/usr/home/bluca
debian/tmp/usr/home/bluca/git
debian/tmp/usr/home/bluca/git/stlink
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib/x86_64-linux-gnu
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib/x86_64-linux-gnu/libstlink.so.1
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib/x86_64-linux-gnu/libstlink.so.1.4.0
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib/x86_64-linux-gnu/libstlink.so
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib/x86_64-linux-gnu/pkgconfig
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib/x86_64-linux-gnu/pkgconfig/stlink.pc
debian/tmp/usr/home/bluca/git/stlink/obj-x86_64-linux-gnu/debian/tmp/usr/lib/x86_64-linux-gnu/libstlink.a
```

Not sure what to change though, sorry

**Comment by Vascom** [comment]
@bluca You set
`-DLIB_INSTALL_DIR=debian/tmp/usr/lib/x86_64-linux-gnu` 
Just change it to
`-DLIB_INSTALL_DIR=/usr/lib/x86_64-linux-gnu/`

**Comment by bluca** [comment]
That works, thanks (as I said, not  CMake expert).

Note that headers have changed path too - that is fine if they are static and arch-independent

**Comment by Vascom** [comment]
@bluca Headers always arch independent and placed to /usr/include/stlink.

**Comment by xor-gate** [comment]
@bluca I understand the frustration of cmake compared to autoconf build system. But it is unmaintainable and hard to write and read for developers, it is easier to understand for packaging people (almost standard build parameters). Are we good now with this PR? What about the defaults without using any build parameters does it work for `make install` ?

**Comment by joede** [comment]
@slyshykO had asked me to test his branch. It doesn't work for me here with Debian 9! IMO the `debian/*.install` files are wrong.

~~~~
   dh_install -O--buildsystem=cmake
dh_install: Cannot find (any matches for) "usr/lib/*/lib*.a" (tried in "." and "debian/tmp")
dh_install: libstlink-dev missing files: usr/lib/*/lib*.a
dh_install: Cannot find (any matches for) "usr/lib/*/pkgconfig/*" (tried in "." and "debian/tmp")
dh_install: libstlink-dev missing files: usr/lib/*/pkgconfig/*
dh_install: Cannot find (any matches for) "usr/lib/*/lib*.so" (tried in "." and "debian/tmp")
dh_install: libstlink-dev missing files: usr/lib/*/lib*.so
dh_install: Cannot find (any matches for) "usr/lib/*/lib*.so.*" (tried in "." and "debian/tmp")
dh_install: libstlink1 missing files: usr/lib/*/lib*.so.*
dh_install: missing files, aborting
debian/rules:14: die Regel für Ziel „binary“ scheiterte
~~~~

**Comment by joede** [comment]
@bluca the removal of the wildcard in `lib/*/` may be fixed by your CMake changes. I'm no CMake expert. What is wrong with the cmake usage for `debian/rules`? And why are the wildcards inside the path needed?

But there are at least two glitches in the `.install` files.

1. as far as I know the pathes must be relative and not absolute! So I removed the leading `/`!
1. the `-dev` package should not have the `.so` file included. This is installed by the library package.

**Comment by bluca** [comment]
@slyshykO could you please also apply this patch together with your changes:

```
--- a/debian/rules
+++ b/debian/rules
@@ -15,5 +15,5 @@ export DEB_BUILD_MAINT_OPTIONS = hardening=+all
 
 override_dh_auto_configure:
        dh_auto_configure -- \
-    -DCMAKE_LIBRARY_PATH=$(DEB_HOST_MULTIARCH) \
+    -DLIB_INSTALL_DIR=/usr/lib/$(DEB_HOST_MULTIARCH) \
     -DSTLINK_UDEV_RULES_DIR='/lib/udev/rules.d'
```

>What is wrong with the cmake usage for debian/rules?

The CMake options are being changed so the CMake parameter used in d/rules doesn't work anymore. The fix is above.

> And why are the wildcards inside the path needed?

Multiarch, see section 9.1.1.4 of the Policy: https://www.debian.org/doc/debian-policy/

> as far as I know the pathes must be relative and not absolute! So I removed the leading /!

dh_install works fine with either, but feel free to change that if you wish

> the -dev package should not have the .so file included. This is installed by the library package.

The unversioned symlink goes into the -dev package, see section 8.4 of the Policy:
https://www.debian.org/doc/debian-policy/#development-files

**Comment by joede** [comment]
@slyshykO you should use `=/usr/lib/$(DEB_HOST_MULTIARCH)` instead of `=$(DEB_HOST_MULTIARCH)`. ;-)

**Comment by joede** [comment]
@bluca after the last change of @slyshykO, I was able to built `.deb` Debian packages from his repository. Looks good so far. ;-)

Is somebody using a Github repository as source for apt? IMO it would be nice to have a source for this new packages. ;-)

**Comment by bluca** [comment]
I've uploaded it to Debian, it will be available after it's been reviewed (no ETA), afterwards I'll upload to stretch-backports as well

**Comment by joede** [comment]
@bluca `stlink` is neither in `testing` nore in `unstable` yet. How can it be uploaded to `stretch-backports`?

**Comment by bluca** [comment]
That's what I meant with "afterwards" - once it's approved and it's in unstable/testing

**Comment by slyshykO** [comment]
@xor-gate  , I think, it is ready for merge.

**Pull request merged**